src/LanguageServer.php

```diff
@@ -157,14 +167,69 @@ private function indexProject(string $rootPath)
 
                 $this->project->loadDocument($uri);
 
+                if ($fileNum % 1000 === 0) {
+                    $this->saveCache();
+                }
+
                 Loop\setTimeout($processFile, 0);
             } else {
                 $duration = (int)(microtime(true) - $startTime);
                 $mem = (int)(memory_get_usage(true) / (1024 * 1024));
                 $this->client->window->logMessage(MessageType::INFO, "All PHP files parsed in $duration seconds. $mem MiB allocated.");
+                $this->saveCache();
             }
         };
 
         Loop\setTimeout($processFile, 0);
     }
+
+    /**
+     * Restores the definition and reference index from the .phpls cache directory, if available
+     *
+     * @return void
+     */
+    public function restoreCache()
```

Comment:
I think this method can return true if both 'definitions' and 'references' are restored correctly. If one of them is missing then cache is invalid.